crates/kitsune_p2p/kitsune_p2p/src/spawn/actor/meta_net_task.rs

```diff
@@ -80,376 +81,37 @@ impl MetaNetTask {
-                                                    }
-                                                }
-                                                wire::Wire::MetricExchange(
-                                                    wire::MetricExchange { space, msgs },
-                                                ) => {
-                                                    let _ = i_s
-                                                        .incoming_metric_exchange(space, msgs)
-                                                        .await;
-                                                }
-                                                wire::Wire::PeerUnsolicited(
-                                                    wire::PeerUnsolicited { peer_list },
-                                                ) => {
-                                                    for peer in peer_list {
-                                                        if let Err(err) = evt_sender
-                                                            .put_agent_info_signed(
-                                                                PutAgentInfoSignedEvt {
-                                                                    space: peer.space.clone(),
-                                                                    peer_data: vec![peer.clone()],
-                                                                },
-                                                            ).await {
-                                                            tracing::warn!(?err, "error processing incoming agent info unsolicited");
-                                                        }
-                                                    }
-                                                }
-                                                wire::Wire::Failure(_)
-                                                | wire::Wire::Call(_)
-                                                | wire::Wire::CallResp(_)
-                                                | wire::Wire::PeerGet(_)
-                                                | wire::Wire::PeerGetResp(_)
-                                                | wire::Wire::PeerQuery(_)
-                                                | wire::Wire::PeerQueryResp(_) => {
-                                                    tracing::warn!(
-                                                        "received non-notify data in a notify"
-                                                    );
-                                                }
-                                            }
-                                        }
-                                    }
-                                }
+                                } => this.handle_notify(url, con, data).await,
+                            } {
+                                shutdown_notify.notify_one();
```

Comment:
The connect and disconnect were already shutting down on fatal errors but request and notify were not. Once I had all the branches of the match returning their errors I was able to move this `shutdown_notify` out of the match and avoid having the same thing 4 times.

It's worth mentioning here that the error handling isn't super clean for the moment. I am confident that fatal errors are being propagated up to the top of this task and shutting it down because there are tests checking that. All other errors are not handled entirely consistently. I think the important ones will get logged but this is an area for future improvement I think